source/common/stats/lazy_init.h

```diff
@@ -0,0 +1,41 @@
+#pragma once
+
+#include "source/common/common/thread.h"
+#include "source/common/runtime/runtime_features.h"
+
+namespace Envoy {
+namespace Stats {
+
+/**
+ * Lazy-initialization wrapper for StatsStructType, intended for deferred instantiation of a block
+ * of stats that might not be needed in a given Envoy process.
+ *
+ * This class is thread-safe -- instantiations can occur on multiple concurrent threads.
+ */
+template <typename StatsStructType> class LazyInit {
+public:
+  // Capture the stat names object and the scope with a ctor, that can be used to instantiate a
+  // StatsStructType object later.
+  // Caller should make sure scope and stat_names outlive this object.
+  LazyInit(Stats::Scope& scope, const typename StatsStructType::StatNameType& stat_names)
+      : ctor_([&scope, &stat_names]() -> StatsStructType* {
+          return new StatsStructType(stat_names, scope);
+        }) {
+    if (Runtime::runtimeFeatureEnabled("envoy.reloadable_features.enable_stats_lazyinit")) {
```

Comment:
I feel like this feature flag will stay around, so we can always not enable the lazy-init behavior, in which case, it works just like before.

Reply:
Generally all the runtime flags get removed after a few release cycles. Things that are supposed to remain forever are part of the config, not the runtime.

Reply:
it's gonna be a lot of plumbing if we decide this needs a config. While the PR is already quite meaty. Can we defer that to a later PR, with a TODO? 

Personally I don't think runtime-flags are always temporary, but I can see your point.

Reply:
I think it's our *policy* that runtime flags are temporary.

However I am not sure we know now whether this feature wants to be permanently config-protected or, once we are confident in it, would we enforce laziness.

My argument for overriding laziness is it might make it harder to know the true memory requirements of Envoy's long-lived structures -- it won't be revealed until traffic occurs. It would be great to hear opinions from the community.

In the meantime, Greg, do you think it's OK to leave as a runtime flag? We can promote to a config option in a follow-up if laziness is not universally accepted.

Reply:
> 

there is a benchmark test I added under test/common/stats:lazyinit_stats_benchmark_test.

There is no problem with accessing the stats (the fourth benchmark), after the nested stats is created. 

There is a performance drop of large scale clusterTrafficStats creations on different workers tho that's not close to the real prod scenario. 

See the following test result: (1 means lazy-init mode) 

benchmarkLazyInitCreation: create lazyinit stats, but not the nested ClusterTrafficStats shows great saving on Main thread.
benchmarkLazyInitCreationInstantiateSameThread: create and instantiate the nested stats showed .5% regress when there are 20K clusters. 
benchmarkLazyInitCreationInstantiateOnWorkerThreads:  create stats on different workers showed performance drop for the lazy-inited stats, but this scenario is not happening in real prod. we'd expect the stats got created gradually as traffic come in. 
benchmarkLazyInitStatsAccess: Create 20K stats in main thread, and inc() one field of the 20K stats 10 times on 5 workers, shows .5% penalty. 


$ bazel ru--compilation_mode=opt --cxxopt=-g --cxxopt=-ggdb3 test/common/stats:lazyinit_stats_benchmark_test

Benchmark                                                            Time             CPU   Iterations
------------------------------------------------------------------------------------------------------
benchmarkLazyInitCreation/0/1000                                  47.0 ms         47.0 ms            1
benchmarkLazyInitCreation/1/1000                                 0.696 ms        0.695 ms            1
benchmarkLazyInitCreation/0/10000                                  582 ms          582 ms            1
benchmarkLazyInitCreation/1/10000                                 8.98 ms         8.98 ms            1
benchmarkLazyInitCreation/0/20000                                 1518 ms         1518 ms            1
benchmarkLazyInitCreation/1/20000                                 17.0 ms         17.0 ms            1
benchmarkLazyInitCreationInstantiateSameThread/0/1000             43.8 ms         43.8 ms            1
benchmarkLazyInitCreationInstantiateSameThread/1/1000             41.5 ms         41.5 ms            1
benchmarkLazyInitCreationInstantiateSameThread/0/10000             595 ms          595 ms            1
benchmarkLazyInitCreationInstantiateSameThread/1/10000             595 ms          595 ms            1
benchmarkLazyInitCreationInstantiateSameThread/0/20000            1293 ms         1293 ms            1
benchmarkLazyInitCreationInstantiateSameThread/1/20000            1300 ms         1300 ms            1
benchmarkLazyInitCreationInstantiateOnWorkerThreads/0/1000         112 ms        0.089 ms            1
benchmarkLazyInitCreationInstantiateOnWorkerThreads/1/1000         469 ms        0.062 ms            1
benchmarkLazyInitCreationInstantiateOnWorkerThreads/0/10000       1067 ms        0.099 ms            1
benchmarkLazyInitCreationInstantiateOnWorkerThreads/1/10000       4758 ms        0.076 ms            1
benchmarkLazyInitCreationInstantiateOnWorkerThreads/0/20000       2167 ms        0.119 ms            1
benchmarkLazyInitCreationInstantiateOnWorkerThreads/1/20000      10674 ms        0.106 ms            1
benchmarkLazyInitStatsAccess/0/1000                               94.9 ms        0.101 ms            1
benchmarkLazyInitStatsAccess/1/1000                                110 ms        0.092 ms            1
benchmarkLazyInitStatsAccess/0/10000                              1057 ms        0.110 ms            1
benchmarkLazyInitStatsAccess/1/10000                              1069 ms        0.128 ms            1
benchmarkLazyInitStatsAccess/0/20000                              2186 ms        0.130 ms            1
benchmarkLazyInitStatsAccess/1/20000                              2199 ms        0.132 ms            1

Reply:
I don't think lazy-created should be enabled by default, so I think we should sort this out now. 

Another issue we may run into: the time-series db (prometheus or whatever) may be fine for "new" envoy instances, but after some traffic lazy-creates various stats it starts to have capacity issues. It's a MUCH simpler system (not just envoy, but the overall system) if the stats are all created at config-load time.

Reply:
Thanks for doing the benchmark. Results look good.